### PR TITLE
Add a CutFEM tutorial as step-85

### PR DESCRIPTION
**Pull request by simonsticko**
This step shows how the NonMatching classes: MeshClassifier, FEImmersedSurfaceValues and NonMatching::FEValues are intended to work together. This is done by solving the Poisson equation on a hypercube mesh where the domain is described by a level set function.

Still somewhat half-written. Opened for discussion. Bump @kronbichler
Depends on #12024, #12101 and #12756 in order to work.

### CODE REVIEW

File: examples/step-85/step-85.cc
Comment (peterrum) [inline examples/step-85/step-85.cc:454]: What about using the MeshWorker infrastructure here (as in step-74)? I am not an expert in that but if I remember correctly it allow you to specify that a face  is visited only once.

File: examples/step-85/step-85.cc
Comment (peterrum) [inline examples/step-85/step-85.cc:510]: Out of curiosity, how big problems can you solve without a preconditioner?

File: examples/step-85/step-85.cc
Comment (peterrum) [inline examples/step-85/step-85.cc:538]: You should be able to add different `DoFHandler`s to the same `DataOut` instance; you just need to call the `add_data_vector` versions that take `DoFHandler` as parameter. If it not working there is probably a bug.

File: examples/step-85/step-85.cc
Comment (peterrum) [inline examples/step-85/step-85.cc:516]:
```suggestion
// Since we use separate DoFHandler instances for the solution and the level set
```

**Review by peterrum — COMMENTED** [review]
Very nice tutorial! I left some comments. I will take a look at the other related open-PRs in the next days. At least I now know how the new classes are intended to be used.

A general remark (partly related to the tutorial and more related to the non-matching stuff): I think part of the user code could be simplified if there would be functions that do not take collections but normal objects. In this tutorial (and in most use cases), we will only need one fe, quadrature, and mapping object; the only place were we need the collection of fes (fe + nothing) is during setup of `DoFHandler`; the rest of the code does not really care, does it?

File: examples/step-85/doc/intro.dox
Comment (peterrum) [inline examples/step-85/doc/intro.dox:40]: Could you add a sentence describing the picture?

File: examples/step-85/doc/intro.dox
Comment (peterrum) [inline examples/step-85/doc/intro.dox:108]: "above method" is not clear. You were introducing the notion of `inside`, `outside`, ... but you referencing what was before.

File: examples/step-85/doc/intro.dox
Comment (peterrum) [inline examples/step-85/doc/intro.dox:142]: What is the meaning of the overline?

File: examples/step-85/doc/intro.dox
Comment (peterrum) [inline examples/step-85/doc/intro.dox:169]:
```suggestion
<h3>Discrete Level Set Function</h3>
```

File: examples/step-85/doc/intro.dox
Comment (peterrum) [inline examples/step-85/doc/intro.dox:231]:
```suggestion
```

File: examples/step-85/doc/intro.dox
Comment (peterrum) [reply examples/step-85/doc/intro.dox:240]: This still needs to be addressed.

File: examples/step-85/step-85.cc
Comment (peterrum) [inline examples/step-85/step-85.cc:19]: Maybe add a sentence here. Take a look at the other tutorials.

File: examples/step-85/step-85.cc
Comment (peterrum) [inline examples/step-85/step-85.cc:66]: Maybe start here a new section? This is not the include section anymore.

File: examples/step-85/step-85.cc
Comment (peterrum) [inline examples/step-85/step-85.cc:100]:
```suggestion
  // We need two separate DoFHandlers. The first manages the DoFs for the
```

File: examples/step-85/doc/intro.dox
Comment (peterrum) [inline examples/step-85/doc/intro.dox:14]:
```suggestion
In this example, we show how to use the cut finite element method (CutFEM) in deal.II.
```

File: examples/step-85/step-85.cc
Comment (peterrum) [inline examples/step-85/step-85.cc:391]:
```suggestion
      // FEFaceValues-like object to integrate over $T \cap \Gamma$. The only thing
```

File: examples/step-85/step-85.cc
Comment (peterrum) [inline examples/step-85/step-85.cc:515]: Maybe only output the relevant cells? You could use `DataOut::set_cell_selection()`.

File: examples/step-85/doc/results.dox
Comment (peterrum) [inline examples/step-85/doc/results.dox:14]: Maybe you could extend the tutorial so that also outputs the L2 norm of the error? For that you could simply use `NonMatching::FEValues`. Also adding a small convergence table here would be nice. It is quite common in the tutorials!

**Review by peterrum — APPROVED** [review]
I left some additional comments. The tutorial looks very good already from my side. 

Lets wait for another review before merging!

**Comment by simonsticko** [comment]
PDF of most recent version: https://www.dropbox.com/s/swjt64vzthd230j/The%20deal.II%20Library%20The%20step-85%20tutorial%20program.pdf?dl=0

File: examples/step-85/doc/intro.dox
Comment (simonsticko) [reply examples/step-85/doc/intro.dox:146]: I don't know if deal.II has a convention for this, but some books define T as an open set (others as a closed set). If T_1 & T_2 are open sets, F is empty. If we add the closure it is correct regardless of how it is defined.

**Review by peterrum — APPROVED** [review]
Thanks!

**Comment by simonsticko** [comment]
It doesn't seem like anyone has more feedback on this. Shouldn't we just merge now? Or do we want to wait more?

**Review by kronbichler — APPROVED** [review]
I think this looks great, thanks for the effort @simonsticko ! I have a few very minor comments.

File: examples/step-85/step-85.cc
Comment (kronbichler) [reply examples/step-85/step-85.cc:326]: Sorry, forget about this one, it seems to work in your PDF, so let's keep the simple form.

File: examples/step-85/doc/intro.dox
Comment (simonsticko) [reply examples/step-85/doc/intro.dox:200]: Added this, but without the part about a different mesh. It has to be the same `Triangulation` right now.

File: examples/step-85/step-85.cc
Comment (simonsticko) [reply examples/step-85/step-85.cc:421]: Added this, but a bit further down.

File: examples/step-85/doc/intro.dox
Comment (kronbichler) [reply examples/step-85/doc/intro.dox:200]: Yes I agree, this is a good compromise. I suspected that the mesh would not be possible at this point, so I thought of rather trying to indicate potentials (but we have no need to oversell things here).

**Review by kronbichler — APPROVED** [review]

**Comment by kronbichler** [comment]
Thanks for the work @simonsticko , this is an excellent tutorial. Let me merge now and postpone possible small clean-ups to follow-up PRs.

**Comment by kronbichler** [comment]
It's great to have these features in deal.II.

**Pull request merged**
